### PR TITLE
[FLINK-15247][Runtime] Wait for all slots to be free before task executor services shutdown upon stopping

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

#10161 introduced canceling and waiting of all tasks in TM to release resources and exit before shutting down TM services on stop. #10034 assigned a separate memory manager per slot and #10330 introduced dynamic slots and state to `TaskSlotTable`. To achieve clean lifecycle of slot, table and TM, this PR introduces freeing of all slots before shutting down TM services on TM stop. Freeing of all slots cancels the running tasks and releases slot resources, including memory manager.
 
## Brief change log

  - Introduce closing future to `TaskSlot`
  - Introduce `TaskSlotTable#closeAsync`
  - Adjust task executor tests
  - Add more test coverage for `TaskSlot` and `TaskSlotTable`

## Verifying this change

CI unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by azagrebin** [comment]
cc @xintongsong

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public class TaskSlotTest {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTest.java]: I think this test class lacks some important tests: Test that the `TaskSlot` is not closed before all contained tasks have terminated. 

Moreover, tests should be in the commit where the feature is being introduced.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTest.java]: We should also test that the `MemoryManager` stays alive until all `Tasks` have terminated.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java [hunk 2]
```diff
@@ -243,8 +252,181 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
...
+		TaskSlotTable taskSlotTable = null;
+		try {
+			taskSlotTable = createTaskSlotTableWithStartedTask(task, new TestingSlotActionsBuilder().build());
+			Iterator<Task> tasks = taskSlotTable.getTasks(jobId);
+			Task nextTask = tasks.next();
+			assertThat(nextTask.getExecutionId(), is(executionAttemptId));
+			assertThat(nextTask.getAllocationId(), is(allocationId));
+			assertThat(tasks.hasNext(), is(false));
+			taskSlotTable.freeSlot(allocationId);
+			taskSlotTable.removeTask(executionAttemptId);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: why are these calls necessary?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java [hunk 3]
```diff
@@ -243,8 +252,181 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
...
+		SlotActions slotActions = new TestingSlotActions(freeSlotFuture::complete, (aid, uid) -> {});
+		Task task = createTask(
+			jobId,
+			executionAttemptId,
+			allocationId,
+			EmptyInvokable.class);
+		TaskSlotTable taskSlotTable = null;
+		try {
+			taskSlotTable = createTaskSlotTableWithStartedTask(task, slotActions);
+			taskSlotTable.freeSlot(allocationId);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: Why do we need to call this method for the test?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java [hunk 4]
```diff
@@ -243,8 +252,181 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
...
+		TaskSlotTable taskSlotTable = null;
+		try {
+			taskSlotTable = createTaskSlotTableWithStartedTask(task, new TestingSlotActionsBuilder().build());
+			TestInterruptableInvokable.STARTED_FUTURE.get();
+			assertThat(taskSlotTable.freeSlot(allocationId), is(-1));
+			TestInterruptableInvokable.INTERRUPTED_FUTURE.get();
+			CompletableFuture<Void> slotFreeFuture = taskSlotTable.freeAllSlots(null);
+			assertThat(slotFreeFuture.isDone(), is(false));
+			TestInterruptableInvokable.DONE_FUTURE.complete(null);
+			taskSlotTable.removeTask(executionAttemptId);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: Why is this necessary?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 5]
```diff
@@ -731,70 +697,6 @@ public void invoke() throws Exception {
 		}
 	}
 
-	@Test
-	public void testTaskInterruptionAndTerminationOnShutdown() throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: By removing this test we lose coverage that we don't stop the `TaskExecutor` before all tasks have been terminated. I couldn't find another test testing this scenario. Hence, please add one or revert the removal.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 6]
```diff
@@ -376,7 +372,7 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 
 		return FutureUtils
 			.runAfterwards(
-				taskCompletionTracker.failIncompleteTasksAndGetTerminationFuture(),
+				taskSlotTable.freeAllSlots(new FlinkException("The task executor is shutting down.")),
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: The current way of ensuring that all tasks are being terminated before the returned future is completed is super hard to understand because the code path is too convoluted and too many components depend on each other to make it happen. It took me actually quite some time to figure this out and I'm somewhat familiar with the code. Instead of introducing additional APIs which need to be called in a very specific manner, I would suggest that we simply call `taskSlotTable.closeAsync()` which returns a future which is completed once all slots are closed. A slot is closed if all its task have terminated. The information about a task being closed does not need to be passed into the `TaskSlotTable` via the `freeSlot` but we could simple take a look at the `Task#getTerminationFuture` in order to control when a `TaskSlot` is completely released.

A good design principle is context independence. If possible a component should work independent of its context meaning that it tries to decouple it self as good as possible from other components. In this case, for example, we require that `TaskExecutor` calls `TaskSlotTable#freeSlot` after we have called `TaskSlotTable#freeAllSlots` and before we can call `close`.

**Comment by tillrohrmann** [comment]
@azagrebin is the PR ready for another round of review?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorSubmissionTest.java [hunk 7]
```diff
@@ -157,6 +159,9 @@ public void testTaskSubmission() throws Exception {
 			tmGateway.submitTask(tdd, env.getJobMasterId(), timeout).get();
 
 			taskRunningFuture.get();
+			taskFinishedFuture.get();
+
+			taskSlotTable.removeTask(tdd.getExecutionAttemptId());
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorSubmissionTest.java]: It was checked in stop method that there is no tasks. Now it is not a problem with new closing approach.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java (hunk 5 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: I returned the test but refactored it to use `StubTaskSlotTable` and checking waiting for `TaskSlotTable.closeAsync` while terminating `TaskExecutor`. Interruption of tasks in `TaskSlotTable.closeAsync` is tested separately.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Indeed, it is complicated. I tried to avoid changing too much the process of freeing slots. At the moment it involves quite some steps:

- JM calls freeSlot and puts the slot into the releasing state
- task update state final (TE)
- remove task (table)
- if slot is in releasing state and it has no running tasks
  - the async table.freeSlot and
  - notifies RM/close JM connection (slot actions impl in TE)

It would be nice that `table.freeSlot` would not have to go through TE but I guess somehow `table.freeSlot and notifies RM/close JM connection` has to be scheduled async into the main thread atomically.

True, it would be nicer if task management was part of the slot as task belongs to it then it could be tested only there and slot freeing/closing would guarantee interruption, now it is mixed into the table class.

The first approach in this PR also kept slot/table not dealing with threading and futures leaving TM taking care of it. I agree that the suggested approach separates better responsibilities w/o changing the existing process of freeing slot too much and introducing async behaviour to slot/table should not be a big problem now.

I implemented the suggested approach with `closeAsync`. Some final cleanup of slot/table (timeService.stop and  memoryManager.shutdown) happens in sync callbacks to the termination of tasks w/o scheduling it to the main thread of TM. Also, the final cleanup of table internal data structures after slot freeing potentially happens after the closing future but better to be scheduled before. Although this does not look implicitly a problem now, it is better to rethink more the design and responsibilities by the next opportunity and keep all actions thread-confined to avoid accidental putting of critical stuff into the callbacks w/o proper synchronisation. As discussed, one solution could be that slot/table get main thread executor in future and become more independent from TM.

**Comment by azagrebin** [comment]
Thanks for the review @tillrohrmann, I addressed comments.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java (hunk 2 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: I removed the unnecessary calls, also for subsequent comments and added comment in code for one where I think it was needed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java [hunk 8]
```diff
@@ -356,7 +373,7 @@ public boolean markSlotActive(AllocationID allocationId) throws SlotNotFoundExce
 	 * @return True if the slot could be marked inactive
 	 */
 	public boolean markSlotInactive(AllocationID allocationId, Time slotTimeout) throws SlotNotFoundException {
-		checkInit();
+		checkStartedOrClosing();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: Why do we allow the closing state here? I think we cannot do this because this might introduce a potential race condition between the `timerService` being stopped and registering a new timeout on it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java [hunk 9]
```diff
@@ -462,7 +471,7 @@ public int freeSlot(AllocationID allocationId, Throwable cause) throws SlotNotFo
 	 * @return True if the timeout is valid; otherwise false
 	 */
 	public boolean isValidTimeout(AllocationID allocationId, UUID ticket) {
-		checkInit();
+		checkStartedOrClosing();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: Same here, this might allow a race condition between this method and the stopping of the `TimerService`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java [hunk 10]
```diff
@@ -605,7 +615,7 @@ public boolean addTask(Task task) throws SlotNotFoundException, SlotNotActiveExc
 	 * @return The removed task if there is any for the given execution attempt id; otherwise null
 	 */
 	public Task removeTask(ExecutionAttemptID executionAttemptID) {
-		checkInit();
+		checkStartedOrClosing();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: Are we sure that `removeTask` can only be called when the state is started or closing? Can't it happen that we call `removeTask` if there is a late `TaskExecutor#updateTaskExecutionState` call?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java [hunk 11]
```diff
@@ -686,7 +696,7 @@ public MemoryManager getTaskMemoryManager(AllocationID allocationID) throws Slot
 
 	@Override
 	public void notifyTimeout(AllocationID key, UUID ticket) {
-		checkInit();
+		checkStartedOrClosing();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: I am not sure whether this always holds true. I could imagine the situation in which the `Timeout` in `TimerService` fires and just before it calls `notifyTimeout`, we finish closing the `TaskSlotTable`. In this situation we would fail this check.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testtasks/TestInterruptableInvokable.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Test invokable which completes the given future when interrupted (can be used only once).
+ */
+public class TestInterruptableInvokable extends AbstractInvokable {
+	private static CompletableFuture<Void> interruptedFuture;
+	private static CompletableFuture<Void> startedFuture;
+	private static CompletableFuture<Void> doneFuture;
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/testtasks/TestInterruptableInvokable.java]: Could be `OneShotLatches` instead of `CompletableFutures`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 13]
```diff
@@ -701,24 +701,22 @@ public void invoke() throws Exception {
...
-		TestInterruptableInvokable.waitForInterrupted();
-
 		// check task executor is waiting for the task completion and has not terminated yet
-		final CompletableFuture<Void> taskExecutorTerminationFuture = submissionContext.taskExecutor.getTerminationFuture();
 		assertThat(taskExecutorTerminationFuture.isDone(), is(false));
 
-		// check task executor has exited after task completion
-		TestInterruptableInvokable.finish();
+		// check task executor has exited after task slot table termination
+		taskSlotTableClosingFuture.complete(null);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Nice work :-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 8 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: As discussed offline, enabling this check would need at least disabling message handling after `RpcEndpoint#onStop` to not accept disconnect messages from job master after `taskExecutor#onStop`. On the other hand, requiring only start method call (not `CREATED`) should be enough because calling it in non-running state will register a timeout in worse case or observe no slot. The issue with concurrent access in `timerService#stop` should be resolved anyways.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 10 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: All `TaskExecutor#updateTaskExecutionState` callbacks with `removeTask` should be scheduled to the main thread in task executor before task termination future is competed. This means that the final `CLOSED` table state will not be set before that. If so, I would keep the check.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 9 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: I think this would not hurt if we relax this check to `STARTED` (not `CREATED`) and make `RUNNING` state check part of the return value. This method is called after `notifyTimeout` where we can also relax it to `STARTED`. Even if we schedule `timerService#stop` to the main thread, `notifyTimeout` can be called afterwards because `timerService#stop` does not block on internal executor shutdown and we do not want it, I guess. If those methods are called afterwards, it should not be problem as they will just find that there is no registered callback anymore.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 11 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: See answer in the comment to `isValidTimeout`

**Comment by azagrebin** [comment]
Thanks for the review @tillrohrmann 
I addressed the comments.

I think the safest way to handle timer service concurrency issues is to schedule its stopping to the main thread of task executor. Then there is only a thread-confined access to its internal container for registered callbacks.

In long term, we should try to make `Task#notifyFinalState` synced with its terminal future because  internal state of slot and table is managed in main thread inside task executor from `notifyFinalState` and the introduced closing is resulted from terminal futures. This leads to some table methods being called after the closed state is reached. This does not look as an immediate problem atm but complicates judging about the components. The same applies to the mentioned separation of responsibilities: TE -> table -> slot -> task.

**Comment by AHeise** [comment]
> I think the safest way to handle timer service concurrency issues is to schedule its stopping to the main thread of task executor. Then there is only a thread-confined access to its internal container for registered callbacks.
Would be awesome if you could use that and simplify the threading model.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java [hunk 14]
```diff
@@ -238,8 +208,112 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
...
+			allocationId,
+			TestInterruptableInvokable.class);
+		final TaskSlotTable taskSlotTable = createTaskSlotTableWithStartedTask(task, new TestingSlotActionsBuilder().build());
+		TestInterruptableInvokable.waitForStarted();
+		assertThat(taskSlotTable.freeSlot(allocationId), is(-1));
+		TestInterruptableInvokable.waitForInterrupted();
+		CompletableFuture<Void> closingFuture = taskSlotTable.closeAsync();
+		assertThat(closingFuture.isDone(), is(false));
+		TestInterruptableInvokable.finish();
+		closingFuture.get();
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: nit: technically we are testing two things here. Don't have to change this now but keep this in mind for the future.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java [hunk 15]
```diff
@@ -238,8 +208,112 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
...
+	}
+
+	private static TaskSlotTableImpl createTaskSlotTableAndStart(int numberOfSlots) {
+		final TaskSlotTableImpl taskSlotTable = TaskSlotUtils.createTaskSlotTable(numberOfSlots);
+		taskSlotTable.start(new TestingSlotActionsBuilder().build(), Executors.directExecutor());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java]: Using the `directExecutor` can be problematic as it allows for thread interleavings which are not possible with a proper `Executor`. I assume that in this case it is not problematic, though, because we use the executor only for the clean up call.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java [hunk 16]
```diff
@@ -73,4 +79,30 @@ public static ResourceProfile createTotalResourceProfile(int numberOfSlots) {
 	public static TimerService<AllocationID> createDefaultTimerService(long shutdownTimeout) {
 		return new TimerService<>(TestingUtils.defaultExecutor(), shutdownTimeout);
 	}
+
+	public static Task createTask1(
+		JobID jobId,
+		ExecutionAttemptID executionAttemptId,
+		AllocationID allocationId,
+		Class<? extends AbstractInvokable> invokable) throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java]: nit: the formatting is off wrt `createTask`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java [hunk 17]
```diff
@@ -73,4 +79,30 @@ public static ResourceProfile createTotalResourceProfile(int numberOfSlots) {
 	public static TimerService<AllocationID> createDefaultTimerService(long shutdownTimeout) {
 		return new TimerService<>(TestingUtils.defaultExecutor(), shutdownTimeout);
 	}
+
+	public static Task createTask1(
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java]: Is this duplicating `createTask`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 18]
```diff
@@ -0,0 +1,777 @@
...
+			closingFuture = FutureUtils
+				.waitForAll(
+					new ArrayList<>(allocatedSlots.values())
+						.stream()
+						.map(slot -> freeSlot(slot, cause))
+						.collect(Collectors.toList()))
+				.thenRunAsync(
+					() -> {
+						state = State.CLOSED;
+						timerService.stop();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java]: Given that this now runs in the main thread, we could also clean up the internal structures.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java (hunk 17 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotUtils.java]: unused leftover

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TestingTaskSlotTable.java [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public SlotReport createSlotReport(ResourceID resourceId) {
+		if (createSlotReportSupplier != null) {
+			return createSlotReportSupplier.get();
+		} else {
+			throw new UnsupportedOperationException();
+		}
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TestingTaskSlotTable.java]: Nit: Instead of this pattern, I would rather set a default `createSlotReportSupplier` in the builder. If possible the default supplier could return something meaningful or it could fail with a more expressive exception.
